### PR TITLE
Clang-tidy some query files

### PR DESCRIPTION
**Pull request by ddn0**

### CODE REVIEW

File: libquery/src/Operator.cpp [hunk 1]
```diff
@@ -1,48 +1,48 @@
...
 
-void Operator::Clear() {
+void galois::Operator::Clear() {
   num_ready_inputs_ = 0;
   std::fill(inputs_.begin(), inputs_.end(), nullptr);
   output_ = nullptr;
 }
 
-std::vector<Operator*> Operator::PropagateOutput() {
+auto galois::Operator::PropagateOutput() -> std::vector<Operator*> {
```
[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/src/Operator.cpp:41]: They are both explicit return types. It is a matter of whether you want to write:

```cpp
std::vector<galois::Operator*> galois::Operator::PropagateOutput() {
```

or 

```cpp
auto galois::Operator::PropagateOutput() -> std::vector<Operator*> {
```

(i.e., how many times names should be prefixed with their namespace)

Once the parser knows the function is in the `galois` namespace, it doesn't need any help with the other types. In the first version, the function is not known to be in the `galois` namespace until after the return value is declared.

File: libquery/src/OpGraph.cpp [hunk 2]
```diff
@@ -31,47 +31,46 @@ void OpGraph::RemoveOp(Operator* op) {
...
+    frontier.clear();
     std::swap(frontier, new_frontier);
   }
 
   GALOIS_LOG_VASSERT(terminal_op != nullptr,
-                     "No terminal op exists in the op graph!");
-  return dynamic_cast<Table*>(terminal_op->GetOutput());
+                     "no terminal op exists in the op graph");
+  auto* ret = dynamic_cast<Table*>(terminal_op->GetOutput());
+  assert(ret != nullptr);
```
[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/src/OpGraph.cpp]:
```
ctest -R sanity -VV
[...]
75: ERROR: /home/ddn/w/katana/libquery/src/OpGraph.cpp:73: assertion not true: ret
1/1 Test #575: executionplantest_sanity .........Child aborted***Exception:  21.27 sec
```

File: libquery/src/Operator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (roshandathathri) [reply libquery/src/Operator.cpp:41]: Thanks!

File: libquery/src/OpGraph.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (roshandathathri) [reply libquery/src/OpGraph.cpp]: @bowu Many queries are tested within a single ctest. Please check the log Testing/Temporary/LastTest.log to know which query causes the issue.

File: libquery/src/Operator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/src/Operator.cpp:41]: FWIW, I don't mind switching back to the original form so there is more consistency between declarations and definitions.

[the remark this answers is not in the document]
Comment (roshandathathri) [reply libquery/src/Operator.cpp:41]: I don't mind either way.

**Comment by ddn0** [comment]
> Maybe we could have a batch mode for clang-tidy like we do for clang-format (scripts/check_format.sh).

I had this originally, but it takes a long time to run and a lot of the existing codebase was written before clang-tidy existed. We can enable it on a per library basis if people want. I think a good starting point would be for people to use clang-tidy manually to get a sense of what it would suggest (and what rules we might disable), and make sure any new files follow its recommendations.

We can follow up with enforcement once the code reaches a higher quality.

Unlike `clang-format`, `clang-tidy -fix` isn't fully automated, so we can't simply turn it on and enforce it immediately.

File: libquery/src/CypherCompiler.cpp [hunk 3]
```diff
@@ -790,6 +794,7 @@ int CypherCompiler::CompileReturnProjection(CypherASTNode projection_ast) {
 }
 
 // contains various helper functions that don't need to access class structures
+namespace galois {
```
[the remark this answers is not in the document]
Comment (l-hoang) [reply libquery/src/CypherCompiler.cpp:797]: Yeah, now that you mention it I should I have defined them outside the namespace.

File: libquery/src/Operator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/src/Operator.cpp]: I've reverted back to the original style since trailing return values is not very common in the codebase.

[the remark this answers is not in the document]
Comment (l-hoang) [reply libquery/src/Operator.cpp]: is this common in other places? I've pretty much never seen this kind of return used outside of maybe one instance.

[the remark this answers is not in the document]
Comment (ddn0) [reply libquery/src/Operator.cpp]: They are common enough to get a linter rule:
https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-trailing-return-type.html

And they present a slightly more uniform style than the traditional leading return type:
https://www.quora.com/When-should-I-use-trailing-return-types-in-C++

Here's a summary of pros and cons: https://blog.petrzemek.net/2017/01/17/pros-and-cons-of-alternative-function-syntax-in-cpp/

TL;DR: most codebases haven't switched to trailing return values (en masse) because it would introduce inconsistency with their existing declarations.
